### DISCUSSION TITLE
Micronaut/Introspection API to search for all interfaces that has a certain annotation.

### DISCUSSION THREAD

**Discussion by subash89** [micronaut-projects/micronaut-core#9439] [Q&A] [answered]
This is not a bean/object but an interface only(a type), annotated with, say DAO.class. I could achieve this in java reflection as follows. Is there a way I can achieve same using introspection/reflection free manner so that I will not have to use reflection. 
     
```
 Reflections reflections = new Reflections("com.abc");
        Set<Class<?>> annotated =
                reflections.getTypesAnnotatedWith(DAO.class);
```

**Comment by yawkat** [comment]
you could try the `scan` methods on Environment.

**Reply by subash89** [reply]
@yawkat  Thanks for the suggestion. 

I tried ctx.getEnvironment().scan(DAO.class,"com.abc") nothing comes up. Idea is to find  'MyBatisTableOne' 


```
@Qualifier
@Retention(RUNTIME)
@Introspected
public @interface DAO {

    String type();

    String datasource();
}
```


And this is the class/interface I want to find. 

```
@DAO(type = "mybatis", datasource = "db1")
@Introspected
public interface MyBatisTableOne{

    @Select("SELECT customer_name FROM TEST_SERVICES WHERE customer_id = #{customer_id}")
    String getOrganizationId(String customer_id);

}
```

**Comment by graemerocher** [comment]
introspections are generally for JavaBeans/records to be able read/write properties.

You can remove `@Introspected` and call `ctx.getBeanDefinitions(Qualifiers.byStereotype(DAO.class))` to find these as beans.

**Reply by subash89** [reply]
@graemerocher  I removed the `@Inrospected` and called this. Still returns me empty list.

**Reply by graemerocher** [reply]
most likely because you have an interface which requires an implementation. You could annotate `DAO` with `@Introduction` although it is a bit weird

**Reply by subash89** [reply]
@graemerocher  Thanks for the clue. I tried it and it gives me the list of classes now. But the problem is it gives me for example the wrapper class micronaut creates. "class com.abc.MyBatisTableOne$Intercepted". Not the real interface itself. I could not find a way to extract the underneath interface.

**Reply by subash89** [reply]
I can do a trick like, trip out the $Intercepted and get the full qualified name and do a Class.forName() to load it. And that gives me the interface obviously. But that again involves reflection. Any better way ?

**Reply by graemerocher** [reply, marked as answer]
you can use `getClass().getInterfaces()`
